File: ocaml/build-in.py
# Needed because ocamlbuild 3.12.1 doesn't support absolute pathnames (4.00.1 does)
# And 4.01.0 fails when given a relative path ("Failure: Pathname.normalize_list: .. is forbidden here.")
import sys
import os
import subprocess

if not hasattr(subprocess, 'check_output'):
    def check_output(*popenargs, **kwargs):
        if 'stdout' in kwargs:
            raise ValueError('stdout argument not allowed, it will be overridden.')
        process = subprocess.Popen(stdout=subprocess.PIPE, *popenargs, **kwargs)
        output, unused_err = process.communicate()
        retcode = process.poll()
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
            raise subprocess.CalledProcessError(retcode, cmd, output=output)
        return output
    subprocess.check_output = check_output

ocaml_version = subprocess.check_output(["ocamlbuild", "-version"], universal_newlines = True).split(' ', 1)[1]
if ocaml_version.startswith('3'):
	from os.path import relpath
	ocaml_build_dir = relpath(sys.argv[1], '.')
else:
	ocaml_build_dir = sys.argv[1]

print("dir", ocaml_build_dir)

# Hack: when we can depend on a full OCaml feed with the build tools, we can remove this.
# Until then, we need to avoid trying to compile against the limited runtime environment.
if 'OCAMLLIB' in os.environ:
	del os.environ['OCAMLLIB']

os.execvp("make", ["make", 'OCAML_BUILDDIR=' + ocaml_build_dir, "ocaml"])

File: ocaml/build-in.ml
(* This is needed because ocamlbuild 3.12.1 doesn't support absolute pathnames (4.00.1 does)
   And 4.01.0 fails when given a relative path ("Failure: Pathname.normalize_list: .. is forbidden here.") *)

#load "str.cma";;
#load "unix.cma";;

let re_OCAMLLIB = Str.regexp_string "OCAMLLIB="

let (|>) f x = x f

let make_relative path =
  if path.[0] = '/' then (
    let cwd = Unix.getcwd () in
    let rel_path = ref (".." ^ path) in
    for i = 1 to String.length cwd - 1 do
      if cwd.[i] = '/' then rel_path := "../" ^ !rel_path
    done;
    !rel_path
  ) else path   (* Already relative *)

let () =
  match Sys.argv with
  | [| _prog; ocaml_build_dir; make_path |] ->
      let ch = Unix.open_process_in "ocamlbuild -version" in
      let ocamlbuild_version = input_line ch in
      let need_relative_path = Str.string_match (Str.regexp "ocamlbuild 3") ocamlbuild_version 0 in

      let ocaml_build_dir =
        if need_relative_path then make_relative ocaml_build_dir
        else ocaml_build_dir in

      (* Hack: when we can depend on a full OCaml feed with the build tools, we can remove this.
         Until then, we need to avoid trying to compile against the limited runtime environment. *)
      let env = Unix.environment ()
      |> Array.to_list
      |> List.filter (fun pair -> not (Str.string_match re_OCAMLLIB pair 0))
      |> Array.of_list in

      Unix.execvpe make_path [| make_path; "OCAML_BUILDDIR=" ^ ocaml_build_dir |] env
  | _ -> failwith "usage: ocaml build-in.ml builddir"
